### PR TITLE
fast path for edge cases times = 0 or 1 or x_size = 1

### LINKED ISSUE #1392: Big speed up: Early return for `vec_rep()` and `vec_rep_each()`
The edge case `times = 1` could be made much more efficient by returning early. Maybe also `times = 0` could benefit a little from an early return. This creates nice speed ups e.g. in `tidyr::expand_grid()`.

## Benchmarks

``` r
library(vctrs)

x <- seq(10e6)
x[[1]] <- 1
vec_rep2 <- function(x, times) {
  if (times == 1L) {
    return(x)
  } else if (times == 0L) {
    vec_ptype(x)
  }
  
  vec_rep(x, times)
}

bench::mark(
  vec_rep = vec_rep(x, 1),
  vec_rep2 = vec_rep2(x, 1),
)
#> # A tibble: 2 x 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 vec_rep      17.5ms     18ms      44.0     114MB     106.
#> 2 vec_rep2      382ns    451ns 1189079.         0B       0

bench::mark(
  vec_rep = vec_rep(x, 0),
  vec_rep2 = vec_rep2(x, 0),
)
#> # A tibble: 2 x 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 vec_rep       819ns    907ns   795080.        0B      0  
#> 2 vec_rep2     1.93µs   2.12µs   396239.        0B     39.6

vec_rep_each2 <- function(x, times) {
  if (times == 1L) {
    return(x)
  } else if (times == 0L) {
    vec_ptype(x)
  }
  
  vec_rep_each(x, times)
}

bench::mark(
  vec_rep_each = vec_rep_each(x, 1),
  vec_rep_each2 = vec_rep_each2(x, 1),
)
#> # A tibble: 2 x 6
#>   expression         min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>    <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 vec_rep_each    25.1ms   25.5ms      39.2     114MB     108.
#> 2 vec_rep_each2    413ns    445ns 1361489.         0B       0

bench::mark(
  vec_rep_each = vec_rep_each(x, 0),
  vec_rep_each2 = vec_rep_each2(x, 0),
)
#> # A tibble: 2 x 6
#>   expression         min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>    <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 vec_rep_each    6.21ms    6.6ms      152.        0B        0
#> 2 vec_rep_each2   6.17ms   6.71ms      149.        0B        0
```

<sup>Created on 2021-06-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

### PR DESCRIPTION
**Pull request by mgirlich**
Fixes #1392 

## Benchmarks
```r
library(vctrs)

x <- 1L + seq(10e6)

bench::mark(
  vec_rep1 = vec_rep(x, 1),
  vec_rep0 = vec_rep(x, 0),
  vec_rep_each1 = vec_rep_each(x, 1),
  vec_rep_each0 = vec_rep_each(x, 0),
  vec_rep_x1 = vec_rep(1L, 100e6),
  rep_x1 = rep(1L, 100e6),
  rep.int_x1 = rep.int(1L, 100e6),
  check = FALSE
)

# This PR
#> Warning: Some expressions had a GC in every iteration; so filtering is disabled.
#> # A tibble: 7 x 13
#>   expression        min  median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result
#>   <bch:expr>    <bch:t> <bch:t>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list>
#> 1 vec_rep1        772ns   882ns 928978.          0B     0    10000     0     10.8ms <NULL>
#> 2 vec_rep0        800ns   890ns 966806.          0B     0    10000     0     10.3ms <NULL>
#> 3 vec_rep_each1   767ns   876ns 982103.          0B     0    10000     0     10.2ms <NULL>
#> 4 vec_rep_each0   796ns   894ns 801876.          0B     0    10000     0     12.5ms <NULL>
#> 5 vec_rep_x1      127ms   134ms      7.31     381MB     7.31     4     4      547ms <NULL>
#> 6 rep_x1          126ms   131ms      7.62     381MB     7.62     4     4    525.2ms <NULL>

# CRAN
#> Warning: Some expressions had a GC in every iteration; so filtering is disabled.
#> # A tibble: 7 x 13
#>   expression         min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
#>   <bch:expr>    <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
#> 1 vec_rep1       11.84ms  14.66ms     44.0     76.3MB    11.5     23     6    522.4ms
#> 2 vec_rep0         834ns    911ns 921936.          0B     0    10000     0     10.8ms
#> 3 vec_rep_each1  19.52ms  21.02ms     29.7     76.3MB     5.93    15     3    505.8ms
#> 4 vec_rep_each0   5.61ms   5.81ms    172.          0B     0       86     0    500.2ms
#> 5 vec_rep_x1    206.72ms 228.49ms      4.46   762.9MB     4.46     3     3      672ms
#> 6 rep_x1         97.12ms 138.09ms      7.40   381.5MB     3.70     4     2    540.8ms
```

### CODE REVIEW

**Comment by lionel-** [comment]
Thanks! Did you check that unit tests cover all cases for which you added branches?

(Stylewise we use sentence-case in commit messages, no need to fix in this PR though)

**Comment by lionel-** [comment]
It looks like the tests are missing a couple of cases: https://app.codecov.io/gh/r-lib/vctrs/compare/1398/diff

**Comment by mgirlich** [comment]
Thanks, I added the test cases.

**Comment by lionel-** [comment]
Thanks again!

File: src/rep.c [hunk 1]
```diff
@@ -86,7 +94,14 @@ static SEXP vec_rep_each(SEXP x, SEXP times) {
 
   if (times_size == 1) {
     const int times_ = r_int_get(times, 0);
-    out = vec_rep_each_uniform(x, times_);
+
+    if (times_ == 1) {
+      out = x;
+    } else if (times_ == 0) {
+      out = vec_ptype(x, args_empty);
```
Comment (DavisVaughan) [inline src/rep.c:101]: @lionel- do you happen to remember if there are any cases where `vec_ptype(x)` would be different from `vec_slice_impl(x, vctrs_shared_empty_int)`? I think the second might be slightly safer here, if there is any chance that they could be different

Comment (lionel-) [reply src/rep.c:101]: Good catch! We're heading towards namedless prototypes, so it would be more correct to slice.

@mgirlich Could you please change this to slicing and add a unit test for the case of a named vector? The result should be empty and named with an empty character vector.

Comment (mgirlich) [reply src/rep.c:101]: Done. I had a quick glance at the source code of `vec_ptype()` and I guess the only special case is `NULL`

``` r
library(vctrs)

vec_rep(NULL, 0)
#> Error: Internal error in `vec_slice_impl()`: Unexpected `NULL`.
vec_ptype(NULL)
#> NULL

# Before and after the PR NULL produces an error in `vec_rep()` and `vec_rep_each()`
vec_rep(NULL, 0)
#> Error: Internal error in `vec_slice_impl()`: Unexpected `NULL`.
vec_rep_each(NULL, 0)
#> Error: Internal error in `vec_slice_impl()`: Unexpected `NULL`.
```

Do you want to handle this? If so, where do you think is the best place?

Comment (lionel-) [reply src/rep.c:101]: I think we might be missing a `vec_assert()` in these functions:

```r
vec_rep(NULL, 2)
#> Error: Internal error in `vec_slice_impl()`: Unexpected `NULL`.
```

We should have an error like this:

```r
vec_slice(NULL, 2)
#> Error: Input must be a vector, not NULL.
```

So I handle this with a vector assertion and some unit tests for the `NULL` cases.

Comment (mgirlich) [reply src/rep.c:101]: I also added `vec_assert()` to `vec_unrep()` as the previous error was not very helpful

```r
vctrs::vec_unrep(NULL)
#> Error: All columns in a tibble must be vectors.
#> x Column `key` is NULL.
```
